# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1465923115 14400
#      Tue Jun 14 12:51:55 2016 -0400
# Node ID 83c824c727c9aafd5ab6b51b9e39a3dc7f4d5512
# Parent  e171851ca3445d7f701b56e4a515744c679f290a
hide gnulib pathlen header

We only needed pathlen for readlink.  Instead of wrapping pathlen,
wrap readlink so that we don't need to specify a buffer size.

* liboctave/wrappers/areadlink-wrapper.c,
liboctave/wrappers/areadlink-wrapper.h: New files.

* liboctave/wrappers/module.mk: Update.

* file-ops.cc: Use new octave_areadlink_wrapper function.
Explicitly include unistd.h.

* changed bootstrap.conf: Use areadlink module instead of readlink.

diff --git a/bootstrap.conf b/bootstrap.conf
--- a/bootstrap.conf
+++ b/bootstrap.conf
@@ -14,16 +14,17 @@
 # GNU General Public License for more details.
 
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 
 # gnulib modules used by this package.
 gnulib_modules="
+  areadlink
   base64
   canonicalize
   chdir
   close
   closedir
   copysign
   copysignf
   crypto/md2
@@ -73,17 +74,16 @@ gnulib_modules="
   nproc
   open
   opendir
   pathmax
   pipe-posix
   progname
   putenv
   readdir
-  readlink
   realloc-gnu
   rename
   rmdir
   round
   roundf
   select
   sigaction
   signal
diff --git a/liboctave/system/file-ops.cc b/liboctave/system/file-ops.cc
--- a/liboctave/system/file-ops.cc
+++ b/liboctave/system/file-ops.cc
@@ -24,39 +24,40 @@ along with Octave; see the file COPYING.
 #  include "config.h"
 #endif
 
 #include <cerrno>
 #include <cstdio>
 #include <cstdlib>
 #include <cstring>
 
+#if (defined (OCTAVE_HAVE_WINDOWS_FILESYSTEM) && ! defined (OCTAVE_HAVE_POSIX_FILESYSTEM))
+#  include <algorithm>
+#endif
+
 #include <iostream>
 #include <vector>
 
 #include <sys/stat.h>
 
-#include "pathmax.h"
+#include <sys/types.h>
+#include <unistd.h>
 
+#include "areadlink-wrapper.h"
 #include "canonicalize-file-name-wrapper.h"
 #include "dir-ops.h"
 #include "file-ops.h"
 #include "file-stat.h"
+#include "gen-tempname-wrapper.h"
 #include "oct-env.h"
 #include "oct-locbuf.h"
 #include "oct-passwd.h"
-#include "pathlen.h"
 #include "quit.h"
 #include "singleton-cleanup.h"
 #include "str-vec.h"
-#include "gen-tempname-wrapper.h"
-
-#if (defined (OCTAVE_HAVE_WINDOWS_FILESYSTEM) && ! defined (OCTAVE_HAVE_POSIX_FILESYSTEM))
-#  include <algorithm>
-#endif
 
 namespace octave
 {
   namespace sys
   {
     file_ops *octave::sys::file_ops::instance = 0;
 
     bool
@@ -504,26 +505,24 @@ namespace octave
 
     int
     readlink (const std::string& path, std::string& result, std::string& msg)
     {
       int status = -1;
 
       msg = "";
 
-      char buf[MAXPATHLEN+1];
+      char *buf = octave_areadlink_wrapper (path.c_str ());
 
-      status = gnulib::readlink (path.c_str (), buf, MAXPATHLEN);
-
-      if (status < 0)
+      if (! buf)
         msg = gnulib::strerror (errno);
       else
         {
-          buf[status] = '\0';
-          result = std::string (buf);
+          result = buf;
+          ::free (buf);
           status = 0;
         }
 
       return status;
     }
 
     int
     rename (const std::string& from, const std::string& to)
diff --git a/liboctave/wrappers/areadlink-wrapper.c b/liboctave/wrappers/areadlink-wrapper.c
new file mode 100644
--- /dev/null
+++ b/liboctave/wrappers/areadlink-wrapper.c
@@ -0,0 +1,41 @@
+/*
+
+Copyright (C) 2016 John W. Eaton
+
+This file is part of Octave.
+
+Octave is free software; you can redistribute it and/or modify it
+under the terms of the GNU General Public License as published by the
+Free Software Foundation; either version 3 of the License, or (at your
+option) any later version.
+
+Octave is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+
+You should have received a copy of the GNU General Public License
+along with Octave; see the file COPYING.  If not, see
+<http://www.gnu.org/licenses/>.
+
+*/
+
+// areadlink is provided by gnulib.  We don't include gnulib headers
+// directly in Octave's C++ source files to avoid problems that may be
+// caused by the way that gnulib overrides standard library functions.
+
+#if defined (HAVE_CONFIG_H)
+#  include "config.h"
+#endif
+
+#include "areadlink.h"
+
+#include "areadlink-wrapper.h"
+
+// Returns pointer to allocated storage.
+
+char *
+octave_areadlink_wrapper (const char *name)
+{
+  return areadlink (name);
+}
diff --git a/liboctave/wrappers/areadlink-wrapper.h b/liboctave/wrappers/areadlink-wrapper.h
new file mode 100644
--- /dev/null
+++ b/liboctave/wrappers/areadlink-wrapper.h
@@ -0,0 +1,37 @@
+/*
+
+Copyright (C) 2016 John W. Eaton
+
+This file is part of Octave.
+
+Octave is free software; you can redistribute it and/or modify it
+under the terms of the GNU General Public License as published by the
+Free Software Foundation; either version 3 of the License, or (at your
+option) any later version.
+
+Octave is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+
+You should have received a copy of the GNU General Public License
+along with Octave; see the file COPYING.  If not, see
+<http://www.gnu.org/licenses/>.
+
+*/
+
+#if ! defined (octave_areadlink_wrapper_h)
+#define octave_areadlink_wrapper_h 1
+
+#if defined __cplusplus
+extern "C" {
+#endif
+
+extern char *
+octave_areadlink_wrapper (const char *name);
+
+#if defined __cplusplus
+}
+#endif
+
+#endif
diff --git a/liboctave/wrappers/module.mk b/liboctave/wrappers/module.mk
--- a/liboctave/wrappers/module.mk
+++ b/liboctave/wrappers/module.mk
@@ -1,9 +1,10 @@
 NOINSTALL_WRAPPERS_INC = \
+  liboctave/wrappers/areadlink-wrapper.h \
   liboctave/wrappers/base64-wrappers.h \
   liboctave/wrappers/canonicalize-file-name-wrapper.h \
   liboctave/wrappers/fpucw-wrapper.h \
   liboctave/wrappers/gen-tempname-wrapper.h \
   liboctave/wrappers/hash-wrappers.h \
   liboctave/wrappers/mkostemp-wrapper.h \
   liboctave/wrappers/nanosleep-wrapper.h \
   liboctave/wrappers/nproc-wrapper.h \
@@ -11,16 +12,17 @@ NOINSTALL_WRAPPERS_INC = \
   liboctave/wrappers/set-program-name-wrapper.h \
   liboctave/wrappers/strftime-wrapper.h \
   liboctave/wrappers/strptime-wrapper.h \
   liboctave/wrappers/uname-wrapper.h \
   liboctave/wrappers/unsetenv-wrapper.h \
   liboctave/wrappers/vasprintf-wrapper.h
 
 WRAPPERS_SRC = \
+  liboctave/wrappers/areadlink-wrapper.c \
   liboctave/wrappers/base64-wrappers.c \
   liboctave/wrappers/canonicalize-file-name-wrapper.c \
   liboctave/wrappers/fpucw-wrapper.c \
   liboctave/wrappers/gen-tempname-wrapper.c \
   liboctave/wrappers/hash-wrappers.c \
   liboctave/wrappers/mkostemp-wrapper.c \
   liboctave/wrappers/nanosleep-wrapper.c \
   liboctave/wrappers/nproc-wrapper.c \
